### DISCUSSION TITLE
Getting repeated transcripts and delayed UtteranceEnd signal

### DISCUSSION THREAD

**Discussion by vickybudhiraja** [deepgram/community#859] [General help] [answered]
### Which Deepgram product feature are you using?

Deepgram API - STT Streaming

### Details

Hi Team,

I was trying to use your awesome lib via Sockets. However, I am facing the following issues and I am kind of stuck, please help:
1- I get duplicate transcripts. For example: If I say, "...what is the weather in Delhi today." I get several JSON responses, with incomplete transcripts and the final ones, there can be several (before `UtteranceEnd`) is a blank one, so never know which one is the correct one!
2- More than sometimes, I do not get the `UtteranceEnd` signal, so my logic never sends it to DeepGram for STT and so the app seems to be hanged!

My dev platform is **Flutter for Android.**

I am using` web_socket_channel: ^3.0.0 `to make socket connections. I am also sending the `KeepAlive` message after every 5 secs.

Any clues, leads, docs, comments & feedback is most welcome!

Sincerely,
Vicky

**Comment by team-deepgram** [comment]
Thanks for asking your question about Deepgram! If you didn't already include it in your post, please be sure to add as much detail as possible so we can assist you efficiently, such as:
- The `request_id` if you have a question about your requests or transcription responses.
- The features you used or the full `api.deepgram.com` URL you sent your request to, including parameters.
- Any code snippets you can share.

**Comment by SandraRodgers** [comment]
Do you have `interim_results` turned on? ([Interim Results documentation](https://developers.deepgram.com/docs/interim-results))

If you do, then the API sends interim results as the audio is being processed. These interim results are partial transcripts that are sent before the final transcript. This is why you're seeing multiple JSON responses with incomplete transcripts. The final transcript is sent when Deepgram detects a pause in the speech or when the audio stream ends. If you're seeing multiple final transcripts (indicated by "speech_final": true), it could be due to pauses in the speech. If you're seeing a blank final transcript, it could be due to noise or silence at the end of the audio stream.

If you're not receiving the UtteranceEnd message, it could be due to the following reasons:
- The `utterance_end_ms` parameter is not set correctly. Make sure it's set to a value of 1000 ms or higher.
- The `interim_results` parameter is not set to true.
- There's continuous background noise that's preventing Deepgram from detecting a pause in the speech.

Also something to check is that you're not KeepAlive as a Data or Stream message. Read more here about how this might cause delays: https://developers.deepgram.com/docs/lower-level-websockets#my-transcription-messages-are-getting-delayed

**Comment by vickybudhiraja** [comment]
Many thanks for your reply. `interim_results=True` is a required param when we want to use `utterance_end_ms` as per the docs: https://developers.deepgram.com/docs/utterance-end

Can you tell me what all params are required to be set, if I want to get the correct `UtteranceEnd` signal?

**Comment by vickybudhiraja** [comment]
Hi Team,

Any update for me? I am stuck here!

Thank you.

Best,
Vicky

**Comment by vickybudhiraja** [comment]
Hello, please is there any update for me?

**Comment by SandraRodgers** [comment, marked as answer]
To correctly receive the UtteranceEnd signal, you need to set the following parameters:
- `utterance_end_ms`: This parameter specifies the number of milliseconds Deepgram should wait before sending the UtteranceEnd message. It's recommended to set this value to 1000 ms or higher.
- `interim_results`: This parameter should be set to true. It enables the sending of interim results, which are preliminary transcripts sent before the final transcript. The UtteranceEnd feature relies on this parameter.
